lib/widgets-customize.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+/**
+ * The sanitization function for incoming values for the `gutenberg_widget_blocks` setting.
+ * It's a JSON string, so it decodes it and encodes it again to make sure it's valid.
+ *
+ * @param string $value The incoming value.
+ */
+function gutenberg_customize_sanitize( $value ) {
+	return json_encode( json_decode( $value ) );
+}
+
+/**
+ * Gutenberg's Customize Register.
+ *
+ * Adds a section to the Customizer for editing widgets with Gutenberg.
+ *
+ * @param \WP_Customize_Manager $wp_customize An instance of the class that controls most of the Theme Customization API for WordPress 3.4 and newer.
+ */
+function gutenberg_widgets_customize_register( $manager ) {
+	require_once __DIR__ . '/class-wp-sidebar-block-editor-control.php';
+
+	$manager->add_setting(
+		'gutenberg_widget_blocks',
+		array(
+			'default'           => '{}',
+			'type'              => 'gutenberg_widget_blocks',
+			'capability'        => 'edit_theme_options',
+			'transport'         => 'postMessage',
+			'sanitize_callback' => 'gutenberg_customize_sanitize',
+		)
+	);
+
+	$manager->add_section(
+		'gutenberg_widget_blocks',
+		array( 'title' => __( 'Widget Blocks', 'gutenberg' ) )
+	);
+	$manager->add_control(
+		new WP_Sidebar_Block_Editor_Control(
+			$manager,
+			'gutenberg_widget_blocks',
+			array(
+				'section'  => 'gutenberg_widget_blocks',
+				'settings' => 'gutenberg_widget_blocks',
+			)
+		)
+	);
+}
+
+/**
+ * Removes the core 'Widgets' panel from the Customizer if block based widgets are enabled.
+ *
+ * @param array $components Core Customizer components list.
+ * @return array (Maybe) modified components list.
+ */
+function gutenberg_remove_widgets_panel( $components ) {
+	if ( ! gutenberg_use_widgets_block_editor() ) {
```

Comment:
We should also check this flag before adding the new setting, section, and control.